cmake/package_tooling.cmake

```diff
@@ -0,0 +1,113 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# The Zephyr package tooling script provides a generic script mode interface
+# to the Zephyr CMake package and module structure.
+#
+# The purpose of this script is to provide a single entry for running any Zephyr
+# build tool that is executed during CMake configure time without creating a
+# complete build system.
+#
+# It does so by loading the Zephyr CMake modules specified with the 'MODULES' or
+# 'USAGE' argument.
+#
+# As basic set of modules are always loaded to ensure identical behavior when
+# running tooling identical to Zephyr CMake configure time.
+#
+# 'USAGE' mode:
+# This mode prints Zephyr build system usage for the given module.
+# Exactly what is printed for the specified module is module specific so consult
+# the documentation for the module itself.
+# For example, USAGE=boards will print a list of boards in the system:
+#   'cmake -DUSAGE=boards -P package_tooling.cmake'
+#
+# 'MODULES' mode:
+# This mode execute the given module identical to Zephyr CMake configure time.
+# When running 'MODULES' mode, then the application source directory must be
+# specified using '-DAPPLICATION_SOURCE_DIR=<path-to-sample>'
+#
+# The build directory will default to current working directory but can be
+# controlled with: '-DAPPLICATION_BINARY_DIR=<path-to-build-folder>'
+#
+# For example, if you were invoking CMake for 'hello_world' sample as:
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> ..
+#
+#   you can invoke only dts module (and dependencies) as:
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> -DAPPLICATION_SOURCE_DIR=.. \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# It is also possible to pass additional build settings.
+# If you invoke CMake for 'hello_world' as:
+#
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> -DOVERLAY_CONFIG=foo.overlay ..
+#
+#   you just add the same argument to tooling like:
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> -DOVERLAY_CONFIG=foo.overlay -DAPPLICATION_SOURCE_DIR=.. \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# Note: the samples CMakeLists.txt file is not processed by package tooling, so
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
```

Comment:
This is an unfortunate limitation, but it makes sense why it has to be this way.

Reply:
I have an idea that I will test out which might remove this limitation.